python/test/unit/geometry/test_bounding_box_tree.py

```diff
@@ -496,3 +497,82 @@ def test_surface_bbtree_collision(dtype):
 
     collisions = compute_collisions_trees(bbtree1, bbtree2)
     assert len(collisions) == 1
+
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_determine_point_ownership(dtype):
+    """Find point owners (ranks and cells) using bounding box trees + global communication
+    and compare to point ownership data results."""
+    comm = MPI.COMM_WORLD
+    rank = comm.Get_rank()
+
+    tdim = 2
+    num_cells_side = 4
+    h = dtype(1.0 / num_cells_side)
+    mesh = create_unit_square(MPI.COMM_WORLD,
+                              num_cells_side, num_cells_side,
+                              CellType.quadrilateral, dtype=dtype)
+    cell_map = mesh.topology.index_map(tdim)
+
+    def test_po_against_bb_tree(mesh,points,cells):
+        cells = np.array([cell for cell in cells if cell < cell_map.size_local])
+        num_points = points.shape[0]
+        po = determine_point_ownership(mesh,points,cells)
+        tree = bb_tree(mesh, mesh.topology.dim, cells)
+        assert(mesh.topology.cell_type in [CellType.quadrilateral,CellType.hexahedron])
+        # Since box elements are considered, bounding box collisions correspond
+        # to the actual collisions
+        tree_col = compute_collisions_points(tree, points)
+        local_owners = np.zeros(2*num_points,dtype=np.int32) # rankwise data before Allreduce
+        owners = local_owners.copy()
+        for ipoint in range(num_points):
+            cell = tree_col.links(ipoint)
+            if len(cell) > 0:
+                local_owners[2*ipoint] = rank
+                local_owners[2*ipoint+1] = cell[0]
+        comm.Allreduce(local_owners, owners, op=MPI.SUM)
+        owner_ranks = owners[[2*i   for i in range(num_points)]]
+        owner_cells = owners[[2*i+1 for i in range(num_points)]]
+
+        assert((po.src_owner() == owner_ranks).all())
+
+        dest_points = po.dest_points()
+        # Since all processors ask about the same points,
+        # dest_points contains repeated data:
+        # dest_points = [p0, p1, p2, p3, ..., pN] * comm.size()
+        ipoint = 1
+        duplicate_start_index = -1
+        while (ipoint < dest_points.shape[0]) and not(duplicate_start_index >= 0):
+            if ((dest_points[ipoint,:] == dest_points[0,:]).all()):
+                duplicate_start_index = ipoint
+            ipoint += 1
```

Comment:
This seems over-engineered and is comparing floating point values with equality rather than `np.allclose`.
As you know that all points are the same on every process, `duplicate_start_index` will always be `num_points`?

Reply:
No, `duplicate_start_index` is the number of points found on a processor. it is smaller than `num_points`. I realize I didn't name this variable correctly.

Example: For a two processor run, both processors ask about the same points eg. `[p0, p1, p2, p3]`. Then if two points are found on each processor, `po.dest_points() = [p0, p1, p0, p1]` for one rank and `po.dest_points() = [p2, p3, p2, p3]` for the other. The duplication stems from both processors asking about the same points.

I will replace the `==` by an `np.allclose()`

Reply:
What about distributing the points at creation, instead of sending in multiple points. As you have all the points computed, you can get a local_range by doing something similar to: https://github.com/FEniCS/dolfinx/blob/06e90b8506d967f1836ffac7f289f49cb0fd17ec/cpp/dolfinx/common/MPI.h#L84-L107 when sending in the points. A Python implementation of this function can be lifted from: https://github.com/jorgensd/adios4dolfinx/blob/8329afed02f5f76bf6cd8d33535437acaf29aca4/src/adios4dolfinx/utils.py#L95-L113

i.e. before this function call
```python
start, stop = compute_local_range(mesh.comm, all_midpoints.shape[0])
test_po_against_bb_tree(mesh, all_midpoints[start:stop])
```